contribs/gnodev/cmd/gnodev/main.go

```diff
@@ -0,0 +1,356 @@
+	fs.StringVar(
+		&c.bindAddr,
+		"web-bind",
+		defaultDevOptions.bindAddr,
+		"verbose output when deving",
+	)
+
+	fs.BoolVar(
+		&c.minimal,
+		"minimal",
+		defaultDevOptions.verbose,
+		"don't load example folder packages along default transaction",
+	)
+
+	fs.BoolVar(
+		&c.verbose,
+		"verbose",
+		defaultDevOptions.verbose,
+		"verbose output when deving",
+	)
+
+	fs.BoolVar(
+		&c.noWatch,
+		"no-watch",
+		defaultDevOptions.noWatch,
+		"watch for files change",
+	)
+
+}
+
+func execDev(cfg *devCfg, args []string, io commands.IO) error {
+	ctx, cancel := context.WithCancelCause(context.Background())
+	defer cancel(nil)
+
+	// guess root dir
+	gnoroot := gnoenv.RootDir()
+
+	pkgpaths, err := parseArgsPackages(io, args)
+	if err != nil {
+		return fmt.Errorf("unable to parse package paths: %w", err)
+	}
+
+	// noopLogger := log.NewTMLogger(log.NewSyncWriter(io.Out))
+	noopLogger := tmlog.NewNopLogger()
+
+	// RAWTerm setup
+	rt := gnodev.NewRawTerm()
+	{
+		restore, err := rt.Init()
+		if err != nil {
+			return fmt.Errorf("unable to init raw term for dev: %s", err)
+		}
+		defer restore()
+
+		// correctly format output for terminal
+		io.SetOut(commands.WriteNopCloser(rt))
+
+		// Setup trap signal
+		osm.TrapSignal(func() {
+			restore()
+			cancel(nil)
+		})
+	}
+
+	nodeOut := rt.NamespacedWriter("Node")
+	webOut := rt.NamespacedWriter("GnoWeb")
+	keyOut := rt.NamespacedWriter("KeyPress")
+
+	var dnode *gnodev.Node
+	{
+		var err error
+		// XXX: redirect node to output file
+		dnode, err = setupDevNode(ctx, noopLogger, cfg, pkgpaths, gnoroot)
+		if err != nil {
+			return err // already formated in setupDevNode
+		}
+		defer dnode.Close()
+	}
+
+	fmt.Fprintf(nodeOut, "Listener: %s\n", dnode.GetRemoteAddress())
+
+	// setup files watcher
+	w, err := setupPkgsWatcher(cfg, dnode.ListPkgs())
+	if err != nil {
+		return fmt.Errorf("unable to watch for files change: %w", err)
+	}
+
+	ccpath := make(chan []string, 1)
+	go func() {
+		defer close(ccpath)
+
+		const debounceTimeout = time.Millisecond * 500
+
+		if err := handleDebounce(ctx, w, ccpath, debounceTimeout); err != nil {
+			cancel(err)
+		}
+	}()
+
+	// Gnoweb setup
+	server := setupGnowebServer(cfg, dnode, rt)
+
+	l, err := net.Listen("tcp", cfg.bindAddr)
+	if err != nil {
+		return fmt.Errorf("unable to listen to %q: %w", cfg.bindAddr, err)
+	}
+
+	go func() {
+		var err error
+		if srvErr := server.Serve(l); srvErr != nil {
+			err = fmt.Errorf("HTTP server stopped with error: %w", srvErr)
+		}
+		cancel(err)
+	}()
+	defer server.Close()
+
+	fmt.Fprintf(webOut, "Listener: %s\n", l.Addr())
+
+	// Print basic infos
+	fmt.Fprintf(nodeOut, "Default Address: %s\n", gnodev.DefaultCreator.String())
+	fmt.Fprintf(nodeOut, "Chain ID: %s\n", dnode.Config().ChainID())
+
+	rt.Taskf("[Ready]", "for commands and help, press `h`")
+
+	cckey := listenForKeyPress(keyOut, rt)
+	for {
+		var err error
+
+		select {
+		case <-ctx.Done():
+			return context.Cause(ctx)
+		case paths := <-ccpath:
```

Comment:
It should not be needed _yet_, because the context should already be canceled if the channel is closed. But you're totally right, I prefer having this case handled anyway.